# E-Commerce Order Data Cleaning & Validation

**Internship:** DecodesLab Virtual Internship  
**Domain:** Data Analytics  
**Tools Used:** Python, Pandas, Jupyter Notebook  
**Dataset Size:** 1200 rows × 14 columns  
**Author:** Ume Habiba

---

## Project Objective
The purpose of this project is to inspect, clean, standardize, and validate an e-commerce orders dataset using Python and Pandas. The workflow focuses on improving data consistency, formatting quality, categorical standardization, and numeric precision while maintaining data integrity.

## Import Libraries

In [2]:
import pandas as pd
import numpy as np


## Load Dataset

In [3]:
df=pd.read_excel('dirty_dataset.xlsx')

In [4]:
df_before=df.copy()


## Inspection Goals
The dataset is inspected to identify:

- Missing values
- Duplicate records
- Inconsistent categorical values
- Incorrect text formatting
- Numeric precision issues
- Invalid or inconsistent date formats

## Data Inspection

In [5]:
df.head()

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10
1,ORD200001,2024-08-23,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,SAVE10,Referral,302.70
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40
3,ORD200003,2023-10-15,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,SAVE10,Facebook,273.19
4,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04


In [6]:
df.tail()

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
1195,ORD201195,2024-06-20,C21126,Desk,1,107.04,392 Main St,Credit Card,Cancelled,TRK38009181,6,FREESHIP,Google,107.04
1196,ORD201196,2024-03-04,C20095,Monitor,2,662.53,778 Main St,Online,Cancelled,TRK69207593,5,NaN,Facebook,1325.06
1197,ORD201197,2023-07-13,C79674,Tablet,2,436.84,275 Main St,Online,Delivered,TRK88039356,2,FREESHIP,Instagram,873.68
1198,ORD201198,2024-08-22,C64753,Chair,4,262.52,509 Main St,Debit Card,Cancelled,TRK71683331,4,WINTER15,Instagram,1050.08
1199,ORD201199,2023-06-11,C57502,Tablet,4,560.58,201 Main St,Gift Card,Returned,TRK51116746,6,SAVE10,Referral,2242.32


In [7]:
df.shape

(1200, 14)

In [8]:
df.columns.tolist()

['OrderID',
 'Date',
 'CustomerID',
 'Product',
 'Quantity',
 'UnitPrice',
 'ShippingAddress',
 'PaymentMethod',
 'OrderStatus',
 'TrackingNumber',
 'ItemsInCart',
 'CouponCode',
 'ReferralSource',
 'TotalPrice']

In [9]:
df.isnull().sum()

OrderID              0
Date                 0
CustomerID           0
Product              0
Quantity             0
UnitPrice            0
ShippingAddress      0
PaymentMethod        0
OrderStatus          0
TrackingNumber       0
ItemsInCart          0
CouponCode         309
ReferralSource       0
TotalPrice           0
dtype: int64

In [10]:
df.dtypes

OrderID                       str
Date               datetime64[us]
CustomerID                    str
Product                       str
Quantity                    int64
UnitPrice                 float64
ShippingAddress               str
PaymentMethod                 str
OrderStatus                   str
TrackingNumber                str
ItemsInCart                 int64
CouponCode                    str
ReferralSource                str
TotalPrice                float64
dtype: object

In [11]:
df['Date'].unique()

<DatetimeArray>
['2023-01-04 00:00:00', '2024-08-23 00:00:00', '2024-02-27 00:00:00',
 '2023-10-15 00:00:00', '2025-05-08 00:00:00', '2023-10-23 00:00:00',
 '2025-06-17 00:00:00', '2023-05-12 00:00:00', '2025-04-02 00:00:00',
 '2023-11-21 00:00:00',
 ...
 '2023-10-17 00:00:00', '2024-12-23 00:00:00', '2024-09-10 00:00:00',
 '2023-04-27 00:00:00', '2024-07-08 00:00:00', '2024-08-17 00:00:00',
 '2025-06-16 00:00:00', '2024-03-04 00:00:00', '2023-07-13 00:00:00',
 '2024-08-22 00:00:00']
Length: 671, dtype: datetime64[us]

In [12]:
print("Full duplicate rows:", df.duplicated().sum())
print("Duplicate OrderIDs:", df['OrderID'].duplicated().sum())

Full duplicate rows: 0
Duplicate OrderIDs: 0


In [13]:
df['OrderStatus'].value_counts()

OrderStatus
Cancelled    250
Returned     247
Pending      237
Shipped      235
Delivered    231
Name: count, dtype: int64

In [14]:
df['ReferralSource'].value_counts()

ReferralSource
Instagram    259
Email        250
Google       241
Facebook     228
Referral     222
Name: count, dtype: int64

In [15]:
df['PaymentMethod'].value_counts()

PaymentMethod
Online         258
Cash           246
Credit Card    234
Debit Card     232
Gift Card      230
Name: count, dtype: int64

In [16]:
print("Quantity <= 0:\n", df[df['Quantity'] <= 0])
print("\nUnitPrice <= 0:\n", df[df['UnitPrice'] <= 0])
print("\nTotalPrice <= 0:\n", df[df['TotalPrice'] <= 0])
print("\nItemsInCart <= 0:\n", df[df['ItemsInCart'] <= 0])

Quantity <= 0:
 Empty DataFrame
Columns: [OrderID, Date, CustomerID, Product, Quantity, UnitPrice, ShippingAddress, PaymentMethod, OrderStatus, TrackingNumber, ItemsInCart, CouponCode, ReferralSource, TotalPrice]
Index: []

UnitPrice <= 0:
 Empty DataFrame
Columns: [OrderID, Date, CustomerID, Product, Quantity, UnitPrice, ShippingAddress, PaymentMethod, OrderStatus, TrackingNumber, ItemsInCart, CouponCode, ReferralSource, TotalPrice]
Index: []

TotalPrice <= 0:
 Empty DataFrame
Columns: [OrderID, Date, CustomerID, Product, Quantity, UnitPrice, ShippingAddress, PaymentMethod, OrderStatus, TrackingNumber, ItemsInCart, CouponCode, ReferralSource, TotalPrice]
Index: []

ItemsInCart <= 0:
 Empty DataFrame
Columns: [OrderID, Date, CustomerID, Product, Quantity, UnitPrice, ShippingAddress, PaymentMethod, OrderStatus, TrackingNumber, ItemsInCart, CouponCode, ReferralSource, TotalPrice]
Index: []


In [17]:
df['calculated'] = df['Quantity'] * df['UnitPrice']
mismatch = df[abs(df['TotalPrice'] - df['calculated']) > 0.01]
print(len(mismatch))
print(mismatch[['OrderID','Quantity','UnitPrice','calculated','TotalPrice']].head(10))

0
Empty DataFrame
Columns: [OrderID, Quantity, UnitPrice, calculated, TotalPrice]
Index: []


In [18]:
df['ShippingAddress'].unique()

<ArrowStringArray>
['928 Main St', '823 Main St', '512 Main St', '275 Main St', '668 Main St',
 '934 Main St', '986 Main St', '706 Main St', '904 Main St', '102 Main St',
 ...
 '105 Main St', '374 Main St', '777 Main St', '676 Main St', '924 Main St',
 '149 Main St', '754 Main St', '392 Main St', '509 Main St', '201 Main St']
Length: 655, dtype: str

In [19]:
df['TrackingNumber'].str.len().value_counts()

TrackingNumber
11    1200
Name: count, dtype: int64

In [20]:
df['TrackingNumber'].head(10)

0    TRK37947903
1    TRK91186779
2    TRK42903982
3    TRK62788070
4    TRK29241424
5    TRK72976927
6    TRK96417362
7    TRK78809193
8    TRK61042692
9    TRK33478363
Name: TrackingNumber, dtype: str

In [21]:
invalid = df[~df['TrackingNumber'].str.match(r'^TRK\d{8}$')]
print(len(invalid))

0



## Inspection Findings

### 1. Basic Info
- Dataset contains 1200 rows and 14 columns
- No structural issues found on initial load

### 2. Duplicates
- No duplicate rows found
- All OrderIDs are unique

### 3. Missing Values
- `CouponCode` — 309 missing values (intentional, not all orders use a coupon)
- All other columns are complete with no missing values

### 4. Numeric Checks
- `Quantity`, `UnitPrice`, `TotalPrice`, `ItemsInCart` — no zero or negative values found

### 5. Categorical Checks
- **PaymentMethod:** `'Online'` is vague — does not specify an actual payment method
- **ReferralSource:** `'Referral'` (222 rows) is vague — no specific source channel mentioned

### 6. Derived Column Validation
- Recalculated `TotalPrice = Quantity × UnitPrice` for all rows
- No mismatches found — data is arithmetically consistent
- Note: CouponCode column exists but no discounts are reflected in TotalPrice

### 7. Address & Tracking
- `ShippingAddress` — synthetic dataset, all addresses contain "Main St"
- `TrackingNumber` — all 1200 values follow valid format: `TRK` + 8 digits

## Data Cleaning

In [26]:
df["ReferralSource"] = df["ReferralSource"].replace({"Referral": "Other"})
df["PaymentMethod"] = df["PaymentMethod"].replace({"Online": "Other"})

In [27]:
df["ReferralSource"].value_counts()

ReferralSource
Instagram    259
Email        250
Google       241
Facebook     228
Other        222
Name: count, dtype: int64

In [28]:
df["PaymentMethod"].value_counts()

PaymentMethod
Other          258
Cash           246
Credit Card    234
Debit Card     232
Gift Card      230
Name: count, dtype: int64

In [29]:
df["Product"] = df["Product"].str.strip().str.title()
df["Product"].value_counts()

Product
Printer    181
Tablet     179
Chair      178
Laptop     173
Desk       170
Monitor    163
Phone      156
Name: count, dtype: int64

In [31]:
df["ShippingAddress"] = df["ShippingAddress"].str.strip().str.title()
df["ShippingAddress"].value_counts()

ShippingAddress
533 Main St    8
895 Main St    6
146 Main St    5
636 Main St    5
562 Main St    5
              ..
149 Main St    1
754 Main St    1
392 Main St    1
509 Main St    1
201 Main St    1
Name: count, Length: 655, dtype: int64

In [32]:
df["ReferralSource"] = df["ReferralSource"].str.strip().str.title()
df["ReferralSource"].value_counts()

ReferralSource
Instagram    259
Email        250
Google       241
Facebook     228
Other        222
Name: count, dtype: int64

### Round numeric columns to 2 decimals

In [33]:
df["UnitPrice"] = df["UnitPrice"].round(2)
df["TotalPrice"] = df["TotalPrice"].round(2)

### Verify Date Format

In [36]:
print(df["Date"].dt.strftime("%Y-%m-%d").head())

0    2023-01-04
1    2024-08-23
2    2024-02-27
3    2023-10-15
4    2025-05-08
Name: Date, dtype: str


## Cleaning Actions Performed
The following preprocessing and cleaning operations were applied:

1. Standardized inconsistent categorical values.
2. Replaced selected categories with unified labels.
3. Applied text normalization using strip and title case formatting.
4. Rounded numeric columns to 2 decimal places.
5. Validated date formatting consistency.
6. Verified dataset integrity after cleaning.

### Export to clean File

In [37]:
df.to_excel("cleaned_dataset.xlsx", index=False)

## Before vs After

| Cleaning Area | Before Cleaning | After Cleaning |
|---|---|---|
| ReferralSource | Contained inconsistent category values | Standardized categories |
| PaymentMethod | Included inconsistent labels | Unified category naming |
| Text Formatting | Mixed casing and extra spaces | Proper title case and trimmed text |
| Numeric Precision | Multiple decimal lengths | Rounded to 2 decimal places |
| Dataset Consistency | Minor formatting inconsistencies | Standardized and analysis-ready |

## Change Log

| Change ID | Description | Impact | Status |
|---|---|---|---|
| CR001 | Replaced 'Referral' in `ReferralSource` with 'Other' | 222 rows standardized | Resolved |
| CR002 | Replaced 'Online' in `PaymentMethod` with 'Other' | Inconsistent category fixed | Resolved |
| CR003 | Applied title case and strip formatting to `Product`, `ShippingAddress`, and `ReferralSource` | Text standardized | Resolved |
| CR004 | Rounded `UnitPrice` and `TotalPrice` to 2 decimal places | Numeric precision enforced | Resolved |

## Conclusion
This project successfully cleaned and validated an e-commerce orders dataset using Python and Pandas.

The workflow included:
- Data inspection
- Missing value analysis
- Duplicate validation
- Category standardization
- Text formatting normalization
- Numeric precision correction
- Final validation and export

The cleaned dataset is now more consistent, reliable, and suitable for further analysis and reporting.